src/validation.cpp

```diff
@@ -2321,7 +2321,7 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
 
             auto res = cache.SubCommunityBalance(CommunityAccountType::IncentiveFunding, distributed);
             if (!res.ok)
-                throw std::runtime_error(strprintf("Pool rewards: can't update community balance: %s. Block %ld (%s)", res.msg, block.height, block.GetHash().ToString()));
+                LogPrintf("Pool rewards: can't update community balance: %s. Block %ld (%s)", res.msg, block.height, block.GetHash().ToString());
```

Comment:
So we still distribute rewards without funding, is that right?

Reply:
Check the line here. From how I see it, the min is chosen, eventually the minimum will be 0 from the community fund and will be the selected amount for block rewards. At that point no rewards will be paid out, but SubCommunityBalance will fail and without the change above would crash the network!

https://github.com/DeFiCh/ain/blob/438ecbb2ef8f368193a1a8ba6914ffaf02deb2ee/src/validation.cpp#L2298

Reply:
If it's 0 will be good case, if it's not, rewards are distributed to all pool pairs without checking is there needed funds. See DistributeRewards, poolReward s distributed to all without checking is there funds for all.

Reply:
It gets the balance for DistributeRewards from the community fund from the line I linked above, basically min of either reward or community fund, so when community fund gets below the reward the remainder is paid out, so the balances is checked before rewards are paid. When the community fund gets to 0 DistributeRewards will have nothing to pay.

Reply:
```
            CAmount poolsBlockReward = std::min(
                                           cache.GetCommunityBalance(CommunityAccountType::IncentiveFunding),
                                           var->dailyReward / (60*60*24/chainparams.GetConsensus().pos.nTargetSpacing) // 2880
                                                );
```
poolsBlockReward  is first argument to DistributeRewards as yieldFarming. The following poolReward is then calculated in DistributeRewards which will be 0 when GetCommunityBalance runs out.
```
CAmount const poolReward = yieldFarming * pool.rewardPct / COIN; 
```

Reply:
And then in `ForEachPoolShare` you reward all pools by this reward `N * poolReward` you don't know it has enough fund for that, no?

Reply:
We do as it uses poolReward which is calculated from above, when 0 it does not run, when less than the DFI reward it pays out whatever is left in the fund.

From ForEachPoolShare:
```
                if (poolReward) {
                    CAmount providerReward = static_cast<CAmount>((arith_uint256(poolReward) * arith_uint256(liquidity) / arith_uint256(pool.totalLiquidity)).GetLow64());
                    if (!newRewardCalc) {
                        providerReward = poolReward * liqWeight / PRECISION;
                    }
                    if (providerReward) {
                        onTransfer(provider, poolId, uint8_t(RewardType::Rewards), {DCT_ID{0}, providerReward}); //can throw
                        totalDistributed += providerReward;
                    }
                }
```

Reply:
We do as it uses poolReward which is calculated from above, when 0 it does not run, when less than the DFI reward it pays out whatever is left in the fund.

From ForEachPoolShare:
```
                if (poolReward) {
                    CAmount providerReward = static_cast<CAmount>((arith_uint256(poolReward) * arith_uint256(liquidity) / arith_uint256(pool.totalLiquidity)).GetLow64());
                    if (!newRewardCalc) {
                        providerReward = poolReward * liqWeight / PRECISION;
                    }
                    if (providerReward) {
                        onTransfer(provider, poolId, uint8_t(RewardType::Rewards), {DCT_ID{0}, providerReward}); //can throw
                        totalDistributed += providerReward;
                    }
                }
```

Reply:
We work out how much each pool gets with...
```
ForEachPoolPair([&] (DCT_ID const & poolId, CPoolPair pool) {
    CAmount const poolReward = yieldFarming * pool.rewardPct / COIN; 
```
Then distribute that per share by weight of liquidity.
```
ForEachPoolShare([&] (DCT_ID const & currentId, CScript const & provider) {
    if (poolReward) {
        CAmount providerReward = static_cast<CAmount>((arith_uint256(poolReward) * arith_uint256(liquidity) / arith_uint256(pool.totalLiquidity)).GetLow64());
```
I'll take a look and check for test this behaves as expected, if there is none then perhaps we should code some to make sure the community fund runs out without a fuss.

Reply:
The problem is liquidity is just provider balance that's not guaranteed community balance has enough funds.